test/common/http/filter/squash_filter_test.cc

```diff
@@ -0,0 +1,487 @@
+    callbacks = nullptr;
+  }
+
+  {
+    // Expect the second get attachment request
+    EXPECT_CALL(cm_.async_client_, send_(_, _, _))
+        .WillOnce(Invoke([&](Envoy::Http::MessagePtr&, Envoy::Http::AsyncClient::Callbacks& cb,
+                             const Envoy::Optional<std::chrono::milliseconds>&)
+                             -> Envoy::Http::AsyncClient::Request* {
+          callbacks = &cb;
+          return &request_;
+        }));
+
+    retry_timer->callback_();
+
+    Http::MessagePtr msg(new Http::ResponseMessageImpl(
+        Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", "200"}}}));
+    msg->body().reset(new Buffer::OwnedImpl(R"EOF({"status":{"state":"attached"}})EOF"));
+
+    EXPECT_CALL(filter_callbacks_, continueDecoding());
+    callbacks->onSuccess(std::move(msg));
+  }
+}
+
+TEST_F(SquashFilterTest, CheckRetryPollingAttachmentOnFailure) {
+  Envoy::Http::AsyncClient::Callbacks* callbacks = startCompleteRequest();
+
+  {
+    // Expect the get attachment request
+    EXPECT_CALL(cm_.async_client_, send_(_, _, _))
+        .WillOnce(Invoke([&](Envoy::Http::MessagePtr&, Envoy::Http::AsyncClient::Callbacks& cb,
+                             const Envoy::Optional<std::chrono::milliseconds>&)
+                             -> Envoy::Http::AsyncClient::Request* {
+          callbacks = &cb;
+          return &request_;
+        }));
+
+    // return the create request
+    Http::MessagePtr msg(new Http::ResponseMessageImpl(
+        Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", "201"}}}));
+    msg->body().reset(new Buffer::OwnedImpl(R"EOF({"metadata":{"name":"a"}})EOF"));
+    callbacks->onSuccess(std::move(msg));
+  }
+  NiceMock<Envoy::Event::MockTimer>* retry_timer;
+  {
+    retry_timer = new NiceMock<Envoy::Event::MockTimer>(&filter_callbacks_.dispatcher_);
+    EXPECT_CALL(*retry_timer, enableTimer(config_->attachment_poll_period()));
+    callbacks->onFailure(Envoy::Http::AsyncClient::FailureReason::Reset);
+    callbacks = nullptr;
+  }
+
+  {
+    // Expect the second get attachment request
+    EXPECT_CALL(cm_.async_client_, send_(_, _, _))
+        .WillOnce(Invoke([&](Envoy::Http::MessagePtr&, Envoy::Http::AsyncClient::Callbacks& cb,
+                             const Envoy::Optional<std::chrono::milliseconds>&)
+                             -> Envoy::Http::AsyncClient::Request* {
+          callbacks = &cb;
+          return &request_;
+        }));
+
+    retry_timer->callback_();
+
+    Http::MessagePtr msg(new Http::ResponseMessageImpl(
+        Http::HeaderMapPtr{new Http::TestHeaderMapImpl{{":status", "200"}}}));
+    msg->body().reset(new Buffer::OwnedImpl(R"EOF({"status":{"state":"attached"}})EOF"));
+
+    EXPECT_CALL(filter_callbacks_, continueDecoding());
+    callbacks->onSuccess(std::move(msg));
+  }
+}
+
+TEST_F(SquashFilterTest, DestroyedInTheMiddle) {
+  Envoy::Http::AsyncClient::Callbacks* callbacks = startCompleteRequest();
+
+  {
```

Comment:
This initial block is repeated several times, can we move it to a common function?